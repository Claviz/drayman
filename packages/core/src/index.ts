import fs from 'fs-extra';
import path from 'path';
import shortid from 'shortid';
import { find, name, path as nodeFindPath } from 'node-find';
import ts from 'typescript';
// import Piscina from 'piscina';
// import { MessageChannel, MessagePort } from 'worker_threads';
import { spawn, Thread, Worker } from "threads"

export const handleComponentEvent = ({ componentInstanceId, eventName, options, files, onSuccess, onError }) => {
    options = options || {};
    if (typeof options === 'string') {
        options = JSON.parse(options);
    }
    const requestId = shortid.generate();
    if (Object.keys(componentInstances).includes(componentInstanceId)) {
        componentInstances[componentInstanceId].eventRequests[requestId] = { onSuccess, onError };
        componentInstances[componentInstanceId].worker.handleComponentEvent({ eventName, requestId, options, files, });
    } else {
        onError({ err: `Component instance was not found.` });
        // res.status(500).send(`Component instance was not found.`);
    }
}

export async function getElementsScriptPaths({ nodeModulesPath = null }) {
    nodeModulesPath = nodeModulesPath || path.join(process.cwd(), 'node_modules');
    const packages = find(nodeFindPath('*/package.json'), { start: nodeModulesPath });
    const paths = {};
    for await (const packageJsonPath of packages) {
        const packageJson = await fs.readJSON(packageJsonPath.toString('/'), { throws: false });
        const elements = packageJson?.drayman?.elements || {};
        for (const element of Object.keys(elements)) {
            paths[element] = path.join(packageJsonPath.parent.toString('/'), elements[element].script);
        }
    }
    return paths;
}

export const onLocationChange = ({ location, connectionId }) => {
    for (const key of Object.keys(componentInstances)) {
        if (componentInstances[key].connectionId === connectionId) {
            componentInstances[key].worker.handleLocationChange({ location });
        }
    }
}

export const onUpdateComponentInstanceProps = ({ componentInstanceId, options, }) => {
    if (typeof options === 'string') {
        options = JSON.parse(options);
    }
    componentInstances[componentInstanceId].worker.updateComponentInstanceProps({ options, });
}

export const onHandleBrowserCallback = ({ callbackId, data, }) => {
    for (const key of Object.keys(componentInstances)) {
        componentInstances[key].worker.handleBrowserCallback({ callbackId, data, });
    }
}

export const handleEventHubEvent = async ({ data, groupId = null, type, namespaceId = null }) => {
    for (const key of Object.keys(componentInstances)) {
        if (componentInstances[key].namespaceId === namespaceId) {
            componentInstances[key].worker.handleEventHubEvent({ type, data, groupId });
        }
    }
}

export const saveComponent = async ({ script, outputFile }) => {

    const tsConfig = JSON.parse(
        await fs.readFile(
            path.join(
                __dirname,
                process.env.NODE_ENV === 'test' ? `../tests/component-processor.tsconfig.test.json` : `../component-processor.tsconfig.json`,
            ),
            'utf-8'
        )
    );
    // const componentOutputDir = path.join(rootDir, `./src/components`);
    // for (const componentFile of componentFiles) {
    // const componentScript = await fs.readFile(path.join(componentInputDir, componentFile), 'utf-8');
    const transpiledComponentScript = ts.transpileModule(script, tsConfig);
    await fs.outputFile(outputFile, transpiledComponentScript.outputText);
    // }
}

// const piscina = new Piscina({
//     filename: path.join(__dirname, `./component-processor.js`),
//     maxThreads: Infinity,
//     // minThreads: 100,
// });

export const onInitializeComponentInstance = async ({
    namespaceId = null,
    extensionsPath = null,
    extensionsOptions = null,
    componentNamePrefix = '',
    componentName,
    componentRootDir,
    componentInstanceId,
    componentOptions,
    location,
    connectionId,
    emit,
    isModal,
    onComponentInstanceConsole,
}) => {

    // const subprocess = execa.node(
    //     path.join(__dirname, `./component-processor.js`),
    //     [
    //         componentInstanceId
    //     ],
    //     { nodeOptions: ['--unhandled-rejections=strict'], serialization: 'advanced', }
    // );
    // const abortController = new AbortController();
    // const { port1, port2 } = new MessageChannel();
    // const { signal } = abortController;
    // port2.on('message', (message) => console.log('received', message));
    // port2.postMessage({ foo: 'bar' });



    // // // subprocess.stdout?.on('data', (data) => {
    // // //     onComponentInstanceConsole?.({ text: data.toString('utf8') });
    // // //     // console.log(data.toString('utf8'));
    // // //     // httpClient.debug({ data: data.toString('utf8') });
    // // // });
    // // // subprocess.stderr?.on('data', (data) => {
    // // //     onComponentInstanceConsole?.({ text: data.toString('utf8') });
    // // //     // console.log(data.toString('utf8'));
    // // //     // httpClient.debug({ data: data.toString('utf8') });
    // // // });
    /**
     * When component instance gets destroyed, we must check for existing user event requests
     * and cancel them.
     */
    // subprocess.on('exit', (x) => {
    //     // const requests = componentInstances[componentInstanceId].eventRequests;
    //     // for (const requestId of Object.keys(requests)) {
    //     //     requests[requestId].status(500).send(`Component instance was destroyed.`);
    //     // }
    //     delete componentInstances[componentInstanceId];
    //     emit({ type: 'componentInstanceDestroyed', payload: {}, componentInstanceId });
    //     // httpClient.sendView({
    //     //     view: null,
    //     // });
    // });
    /**
     * Handles messages sent by component instance process.
     */
    const worker = await spawn(new Worker(`./component-processor.js`));
    componentInstances[componentInstanceId] = {
        // abortController,
        // messagePort: port2,
        worker,
        terminate: async () => {
            await Thread.terminate(worker);
            delete componentInstances[componentInstanceId];
            emit({ type: 'componentInstanceDestroyed', payload: {}, componentInstanceId });
        },
        // process: subprocess,
        /**
         * Used to store user event requests (on button click, on input, etc.).
         */
        eventRequests: {},
        connectionId,
        namespaceId,
    };
    if (toTerminate[componentInstanceId]) {
        componentInstances[componentInstanceId].terminate();
        delete toTerminate[componentInstanceId];
    } else {
        worker.events().subscribe(({ type, payload }) => {
            /**
             * When there is a response to user event request.
             */
            if (type === 'response') {
                const { requestId, result, err } = payload;
                if (err) {
                    // emit({ type: 'componentEventError', payload: { err, requestId }, componentInstanceId });
                    componentInstances[componentInstanceId].eventRequests[requestId].onError({ err });
                    // componentInstances[componentInstanceId].eventRequests[requestId].status(500).send(`${err}`);
                } else {
                    // emit({ type: 'componentEventSuccess', payload: { result, requestId }, componentInstanceId });
                    componentInstances[componentInstanceId].eventRequests[requestId].onSuccess({ result });
                    // componentInstances[componentInstanceId].eventRequests[requestId].json(result || null);
                }
                delete componentInstances[componentInstanceId].eventRequests[requestId];
            } else if (type === 'eventHubEvent') {
                const { eventPayload, groupId, type } = payload;
                handleEventHubEvent({ data: eventPayload, groupId, type, namespaceId });
            } else if (type === 'console') {
                const { text } = payload;
                onComponentInstanceConsole?.({ text });
            } else {
                emit({ type, payload, componentInstanceId });
            }
        })
        worker.initializeComponentInstance({ componentNamePrefix, componentName, componentRootDir, componentOptions, componentInstanceId, location, extensionsPath, extensionsOptions, isModal });
    }

    // port2.postMessage({ type: 'init', payload: { componentNamePrefix, componentName, componentRootDir, componentOptions, componentInstanceId, location, extensionsPath, extensionsOptions, isModal } });
    // Thread.
    // piscina.run({
    //     port: port1,
    //     initData: { componentNamePrefix, componentName, componentRootDir, componentOptions, componentInstanceId, location, extensionsPath, extensionsOptions, isModal }
    // }, { transferList: [port1], signal, }).catch(e => {
    //     delete componentInstances[componentInstanceId];
    //     emit({ type: 'componentInstanceDestroyed', payload: {}, componentInstanceId });
    // });
}

export const onDisconnect = ({ connectionId }) => {
    for (const componentInstanceId of Object.keys(componentInstances).filter(x => componentInstances[x].connectionId === connectionId)) {
        onDestroyComponentInstance({ componentInstanceId });
    }
}

const toTerminate = {};

export const onDestroyComponentInstance = ({ componentInstanceId }) => {
    if (!componentInstances[componentInstanceId]) {
        toTerminate[componentInstanceId] = true;
    } else {
        componentInstances[componentInstanceId].terminate();
    }
    //     delete componentInstances[componentInstanceId];
    //     emit({ type: 'componentInstanceDestroyed', payload: {}, componentInstanceId });
}

export const componentInstances: {
    [componentInstanceId: string]: {
        // messagePort: MessagePort;
        // abortController: AbortController;
        terminate: () => Promise<void>;
        worker: any;
        eventRequests: {
            [requestId: string]: {
                onSuccess: (data: { result: any }) => void,
                onError: (data: { err: any }) => void,
            };
        };
        connectionId: string;
        namespaceId: string;
    }
} = {};
