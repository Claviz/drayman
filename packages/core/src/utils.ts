import { svgElementAttributes, htmlElementAttributes } from 'all-html-attributes';

export const isEvent = (optionName: string) => optionName?.length > 2 && optionName.slice(0, 2) === 'on';
// export const isEvent = (optionName: string) => optionName?.length > 2 && optionName.slice(0, 2) === 'on' && optionName[2] === optionName[2].toUpperCase();

async function parseFnType(item) {
    if (typeof item?.type === 'function') {
        const rendered = await item.type(item.props);
        return parseFnType(rendered);
    }
    return item;
}

export const render = async (raw, childRenderer = null, renderedArr = [], events = {}, parent = '') => {
    async function traverse(arr: { type: string | any; props: any; key: string; }[], renderedArr = [], parent = null) {
        for (let item of (Array.isArray(arr) ? arr : [arr])) {
            // console.log(1, item)
            if (item?.type === '$$fragment') {
                await traverse(item.props?.children, renderedArr, parent);
            } else if (item?.type) {
                // if (typeof item.type === 'function') {
                //     const rendered = await item.type(item.props);
                //     item = { ...rendered };
                // }
                item = { ...await parseFnType(item) }
                const result = await childRenderer?.(item.type, item.key ? `${parent}/${item.key}/${item.type}` : `${parent}/${renderedArr.length}/${item.type}`, item.props);
                if (result) {
                    await render(result, childRenderer, renderedArr, events, `${parent}`);
                } else {
                    const type = (item.type === 'container' || item.type === 'html') ? 'div' : item.type;
                    const { children, style, ...options } = item.props || {};
                    // const options = rest || {};
                    const itemName = `${parent}/${renderedArr.length}/${type}`;
                    const on = {};
                    const props = {};
                    const attrs = {};
                    for (const optionKey of Object.keys(options)) {
                        if (isEvent(optionKey)) {
                            // const eventId = Object.keys(events).length;
                            let eventOptions;
                            let eventFn;
                            if (typeof options[optionKey] === 'function') {
                                eventFn = options[optionKey];
                                eventOptions = true;
                            } else {
                                [eventFn, eventOptions] = options[optionKey];
                            }
                            if ((type as string).includes('-')) {
                                events[`${itemName}/${optionKey}`] = eventFn;
                                props[optionKey] = eventOptions;
                            } else {
                                const eventName = optionKey.slice(2).toLowerCase();
                                events[`${itemName}/${eventName}`] = eventFn;
                                on[eventName] = eventOptions;
                                // delete options[optionKey];
                            }
                            // options[optionKey] = true;
                        } else if (optionKey === 'ref' || (svgElementAttributes[type] && (svgElementAttributes[type].includes(optionKey) || svgElementAttributes['*'].includes(optionKey)))) {
                            attrs[optionKey] = options[optionKey];
                        } else {
                            if (htmlElementAttributes[type] && (htmlElementAttributes[type].includes(optionKey) || htmlElementAttributes['*'].includes(optionKey))) {
                                attrs[optionKey] = options[optionKey];
                            }
                            props[optionKey] = options[optionKey];
                        }
                    }
                    const newItem: any = {
                        // key: itemName,
                        children: [],
                        sel: type,
                        key: itemName,
                        data: {},
                        // data: { props: options },
                        //text: undefined,
                        // options: options,
                        // style: formatStyle(style) || null,
                        // type,
                        // children: [],
                        // key: itemName,
                    };
                    if (style && Object.keys(style).length) {
                        newItem.data.style = formatStyle(style);
                    }
                    if (Object.keys(props).length) {
                        newItem.data.props = props;
                    }
                    if (Object.keys(attrs).length) {
                        newItem.data.attrs = attrs;
                    }
                    if (Object.keys(on).length) {
                        newItem.data.on = on;
                    }
                    if (options.is) {
                        newItem.data.is = options.is;
                    }
                    renderedArr.push(newItem);
                    // if (parentEl?.sel === 'select') {
                    //     if (parentEl.data.props.multiple && parentEl.data.props.value.includes(newItem.data.props.value)) {
                    //         newItem.data.props.selected = true;
                    //     } else if (parentEl.data.props.value === newItem.data.props.value) {
                    //         newItem.data.props.selected = true;
                    //     }
                    // }
                    // if (typeof item.props.children === 'string' || typeof item.props.children === 'number') {
                    //     newItem.text = item.props.children;
                    // } else {
                    await traverse(children, newItem.children, itemName);
                    // if (newItem.sel === 'select' && newItem.data.props?.value) {
                    //     delete newItem.data.props.value;
                    // }
                    // if (newItem.sel === 'input' && newItem.data.props?.type === 'checkbox') {
                    //     newItem.data.props.checked = !!newItem.data.props.value;
                    //     delete newItem.data.props.value;
                    // }
                    // if (newItem.sel === 'input' && newItem.data.props?.type === 'radio') {
                    //     newItem.data.props.checked = !!newItem.data.props.value;
                    //     // delete newItem.data.props.value;
                    // }
                    // if (newItem.sel === 'input' && newItem.data.props?.value) {
                    //     newItem.data.attrs = { value: newItem.data.props.value }
                    //     delete newItem.data.props.value;
                    // }
                    // }
                }
            } else if (Array.isArray(item)) {
                await traverse(item, renderedArr, `${parent}`);
            } else if (item && typeof item === 'string' || typeof item === 'number') {
                // parentEl.text = parentEl.text ? parentEl.text + item : item;
                const lastItem = renderedArr[renderedArr.length - 1];
                if (lastItem && !lastItem.sel) {
                    lastItem.text += item;
                } else {
                    renderedArr.push({
                        children: [],
                        text: item,
                        // sel: type,
                        key: `${parent}/${renderedArr.length}/text`,
                        // data: { props: options },
                        //text: undefined,
                        // options: options,
                        // style: formatStyle(style) || null,
                        // type,
                        // children: [],
                        // key: itemName,
                    })
                }
                // console.log(item)
                // const lastArrItem = renderedArr[renderedArr.length - 1];
                // console.log({ item })
                // console.log({ lastArrItem, item })
                // lastArrItem.text += item;
                // if (lastArrItem?.type === 'text') {
                //     lastArrItem.text += item;
                // } else {
                //     renderedArr.push({
                //         options: { text: item },
                //         // style: null,
                //         // type: 'text',
                //         children: [],
                //         key: `${parent}/${renderedArr.length}/text`,
                //     });
                // }
            }
            // if (item?.props?.children) {
            //     await traverse(item.props.children, renderedArr);
            // }
        }
    }
    await traverse(raw, renderedArr, parent);

    return { tree: renderedArr, events };
}

const formatStyle = (style) => {
    if (!style) {
        return null;
    }
    const newStyle = {};
    for (const prop of Object.keys(style || {})) {
        if (typeof style[prop] === 'number' && !isUnitlessNumber[prop]) {
            style[prop] = `${style[prop]}px`;
        }
        newStyle[prop] = style[prop];
    }
    return newStyle;
}

const isUnitlessNumber = {
    animationIterationCount: true,
    borderImageOutset: true,
    borderImageSlice: true,
    borderImageWidth: true,
    boxFlex: true,
    boxFlexGroup: true,
    boxOrdinalGroup: true,
    columnCount: true,
    columns: true,
    flex: true,
    flexGrow: true,
    flexPositive: true,
    flexShrink: true,
    flexNegative: true,
    flexOrder: true,
    gridRow: true,
    gridRowEnd: true,
    gridRowSpan: true,
    gridRowStart: true,
    gridColumn: true,
    gridColumnEnd: true,
    gridColumnSpan: true,
    gridColumnStart: true,
    fontWeight: true,
    lineClamp: true,
    lineHeight: true,
    opacity: true,
    order: true,
    orphans: true,
    tabSize: true,
    widows: true,
    zIndex: true,
    zoom: true,
    fillOpacity: true,
    floodOpacity: true,
    stopOpacity: true,
    strokeDasharray: true,
    strokeDashoffset: true,
    strokeMiterlimit: true,
    strokeOpacity: true,
    strokeWidth: true,
};

const toKebabCase = (string) => {
    return string.replace(/\B(?:([A-Z])(?=[a-z]))|(?:(?<=[a-z0-9])([A-Z]))/g, '-$1$2').toLowerCase();
}