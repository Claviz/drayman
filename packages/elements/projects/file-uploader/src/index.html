<!doctype html>
<html lang="en">

<head>
  <meta charset="utf-8">
  <title>FileUploader</title>
  <base href="/">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link rel="icon" type="image/x-icon" href="favicon.ico">
</head>

<body>
  <drayman-file-uploader></drayman-file-uploader>
</body>

</html>