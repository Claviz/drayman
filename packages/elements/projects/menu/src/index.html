<!doctype html>
<html lang="en">

<head>
  <meta charset="utf-8">
  <title>Menu</title>
  <base href="/">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link rel="icon" type="image/x-icon" href="favicon.ico">
  <script src="https://unpkg.com/drayman-component-viewer"></script>
</head>

<body>
  <script>
    createComponent('drayman-menu');
  </script>
</body>

</html>