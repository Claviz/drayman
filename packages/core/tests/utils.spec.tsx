import { isEvent, render } from '../src/utils';

test('renders single element', async () => {
    const result = await render(
        <button style={{ color: 'red' }}>Click me!</button>
    );

    expect(result).toMatchSnapshot();
});

test('bindings are working', async () => {
    const text = 'world';
    const result = await render(
        <div>
            <div>Hello, {text}!</div>
        </div>
    );

    expect(result).toMatchSnapshot();
});

test('sets SVG attributes', async () => {
    const result = await render(
        <svg width="100" height="100">
            <circle cx="50" cy="50" r="20" stroke="green" stroke-width="4" fill="yellow" />
        </svg>
    );

    expect(result).toMatchSnapshot();
});

test(`renders elements mixed with text`, async () => {
    const result = await render(
        <div>
            Hello,
            <button>world</button>
            !
        </div>
    );

    expect(result).toMatchSnapshot();
});

test(`renders nested elements`, async () => {
    const result = await render(
        <div>
            <button style={{ color: 'red' }}>Red button</button>
            <button style={{ color: 'blue' }}>Blue button</button>
        </div>
    );

    expect(result).toMatchSnapshot();
});

test(`ignores empty elements`, async () => {
    const result = await render(
        <div>
            {undefined}
            {null}
            <h1>Hello, world!</h1>
            {true}
            {false}
            {''}
        </div>
    );
    expect(result).toMatchSnapshot();
});

test(`function as children`, async () => {
    function Repeat(props) {
        let items = [];
        for (let i = 0; i < props.numTimes; i++) {
            items.push(props.children(i));
        }
        return <div>{items}</div>;
    }
    const result = await render(
        <Repeat numTimes={3}>
            {(index) => <p>This is item {index} in the list</p>}
        </Repeat>
    );

    expect(result).toMatchSnapshot();
});

test(`expression as children 2`, async () => {
    const items = ['This is item 0 in the list', 'This is item 1 in the list', 'This is item 2 in the list'];
    function SubItem({ message }) {
        return <p>{message}</p>;
    }
    function Item({ message }) {
        return <SubItem message={message} />;
    }
    const result = await render(
        <div>
            {items.map((message) => <Item message={message} />)}
        </div>
    );

    expect(result).toMatchSnapshot();
});

test(`detects events`, async () => {
    expect(isEvent('onClick')).toBeTruthy();
    expect(isEvent('onA')).toBeTruthy();
    expect(isEvent('onclick')).toBeTruthy();
    // expect(isEvent('ona')).toBeFalsy();
    // expect(isEvent('one')).toBeFalsy();
    // expect(isEvent('on')).toBeFalsy();
    expect(isEvent('o')).toBeFalsy();
    expect(isEvent('oNClick')).toBeFalsy();
    expect(isEvent('ONClick')).toBeFalsy();
});

test(`renders components with childRenderer`, async () => {
    const renderedChildKeys = [];

    function Child({ message }) {
        return <child message={message} />;
    }

    const childRenderer = async (type, key, props) => {
        if (type === 'child') {
            renderedChildKeys.push(key);

            return (
                <h3>
                    {props.message}
                </h3>
            );
        } else if (type === 'other-child') {
            renderedChildKeys.push(key);

            return (
                <h6>
                    {props.message}
                </h6>
            );
        }

        return null;
    }

    const result = await render(
        [
            <div>
                <child message="Hello, world!" />
                <not-child message="Hello, world!" />
            </div>,
            <child message="Hello, orphan!" />,
            <Child message="Hello, func orphan!" />,
            <other-child message="Hello, other child!" />,
        ],
        childRenderer
    )

    expect(renderedChildKeys).toEqual([
        '/0/div/0/child',
        '/1/child',
        '/2/child',
        '/3/other-child'
    ]);

    expect(result).toMatchSnapshot();
});

test(`renders components with childRenderer and custom key`, async () => {
    const renderedChildKeys = [];
    const childRenderer = async (type, key, props) => {
        if (type === 'child') {
            renderedChildKeys.push(key);

            return (
                <h3>
                    {props.message}
                </h3>
            );
        }
    }
    const result = await render(
        [
            <child key="custom-1" message="1"></child>,
            <child key="custom-2" message="2"></child>,
            <child key="custom-3" message="3"></child>,
        ],
        childRenderer,
    )
    expect(renderedChildKeys).toEqual([
        '/custom-1/child',
        '/custom-2/child',
        '/custom-3/child'
    ]);
    expect(result).toMatchSnapshot();
});

test('returns options for events', async () => {
    const result = await render(
        <input type="text" onInput={[async () => { }, { debounce: 500 }]} />
    );

    expect(result).toMatchSnapshot();
});

test('stores events correctly', async () => {
    const result = await render(
        <input type="text" onInput={async () => { }} />
    );

    expect(result).toMatchSnapshot();
});

test('stores events for custom elements correctly', async () => {
    const result = await render(
        <custom-element onInput={async () => { }} />
    );

    expect(result).toMatchSnapshot();
});

test(`renders nested array of elements`, async () => {
    const result = await render(
        [
            <h1>Hello, world!</h1>,
            [
                <h1>Hello, world!</h1>,
            ]
        ]
    );
    expect(result).toMatchSnapshot();
});

test(`renders fragments as nested array of elements`, async () => {
    const result = await render(
        <>
            <h1>Hello, world!</h1>
            <>
                <h1>Hello, world!</h1>
            </>
        </>
    );
    expect(result).toMatchSnapshot();
});

test(`renders styles appending 'px' to number values where possible`, async () => {
    const result = await render(
        <div style={{ width: 100, height: 50, flex: 1 }}>Hello, world!</div>
    );
    expect(result).toMatchSnapshot();
});

test(`custom CSS properties are allowed`, async () => {
    const result = await render(
        <div style={{ ['--theme-color' as any]: 'black' }}>Hello, world!</div>
    );
    expect(result).toMatchSnapshot();
});