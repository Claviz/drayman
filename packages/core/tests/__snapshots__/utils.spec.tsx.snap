// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bindings are working 1`] = `
Object {
  "events": Object {},
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "/0/div/0/div/0/text",
              "text": "Hello, world!",
            },
          ],
          "data": Object {},
          "key": "/0/div/0/div",
          "sel": "div",
        },
      ],
      "data": Object {},
      "key": "/0/div",
      "sel": "div",
    },
  ],
}
`;

exports[`custom CSS properties are allowed 1`] = `
Object {
  "events": Object {},
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "key": "/0/div/0/text",
          "text": "Hello, world!",
        },
      ],
      "data": Object {
        "style": Object {
          "--theme-color": "black",
        },
      },
      "key": "/0/div",
      "sel": "div",
    },
  ],
}
`;

exports[`expression as children 2 1`] = `
Object {
  "events": Object {},
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "/0/div/0/p/0/text",
              "text": "This is item 0 in the list",
            },
          ],
          "data": Object {},
          "key": "/0/div/0/p",
          "sel": "p",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "/0/div/1/p/0/text",
              "text": "This is item 1 in the list",
            },
          ],
          "data": Object {},
          "key": "/0/div/1/p",
          "sel": "p",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "/0/div/2/p/0/text",
              "text": "This is item 2 in the list",
            },
          ],
          "data": Object {},
          "key": "/0/div/2/p",
          "sel": "p",
        },
      ],
      "data": Object {},
      "key": "/0/div",
      "sel": "div",
    },
  ],
}
`;

exports[`function as children 1`] = `
Object {
  "events": Object {},
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "/0/div/0/p/0/text",
              "text": "This is item 0 in the list",
            },
          ],
          "data": Object {},
          "key": "/0/div/0/p",
          "sel": "p",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "/0/div/1/p/0/text",
              "text": "This is item 1 in the list",
            },
          ],
          "data": Object {},
          "key": "/0/div/1/p",
          "sel": "p",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "/0/div/2/p/0/text",
              "text": "This is item 2 in the list",
            },
          ],
          "data": Object {},
          "key": "/0/div/2/p",
          "sel": "p",
        },
      ],
      "data": Object {},
      "key": "/0/div",
      "sel": "div",
    },
  ],
}
`;

exports[`ignores empty elements 1`] = `
Object {
  "events": Object {},
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "/0/div/0/h1/0/text",
              "text": "Hello, world!",
            },
          ],
          "data": Object {},
          "key": "/0/div/0/h1",
          "sel": "h1",
        },
      ],
      "data": Object {},
      "key": "/0/div",
      "sel": "div",
    },
  ],
}
`;

exports[`renders components with childRenderer 1`] = `
Object {
  "events": Object {},
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "/0/div/0/h3/0/text",
              "text": "Hello, world!",
            },
          ],
          "data": Object {},
          "key": "/0/div/0/h3",
          "sel": "h3",
        },
        Object {
          "children": Array [],
          "data": Object {
            "props": Object {
              "message": "Hello, world!",
            },
          },
          "key": "/0/div/1/not-child",
          "sel": "not-child",
        },
      ],
      "data": Object {},
      "key": "/0/div",
      "sel": "div",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "key": "/1/h3/0/text",
          "text": "Hello, orphan!",
        },
      ],
      "data": Object {},
      "key": "/1/h3",
      "sel": "h3",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "key": "/2/h3/0/text",
          "text": "Hello, func orphan!",
        },
      ],
      "data": Object {},
      "key": "/2/h3",
      "sel": "h3",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "key": "/3/h6/0/text",
          "text": "Hello, other child!",
        },
      ],
      "data": Object {},
      "key": "/3/h6",
      "sel": "h6",
    },
  ],
}
`;

exports[`renders components with childRenderer and custom key 1`] = `
Object {
  "events": Object {},
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "key": "/0/h3/0/text",
          "text": "1",
        },
      ],
      "data": Object {},
      "key": "/0/h3",
      "sel": "h3",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "key": "/1/h3/0/text",
          "text": "2",
        },
      ],
      "data": Object {},
      "key": "/1/h3",
      "sel": "h3",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "key": "/2/h3/0/text",
          "text": "3",
        },
      ],
      "data": Object {},
      "key": "/2/h3",
      "sel": "h3",
    },
  ],
}
`;

exports[`renders elements mixed with text 1`] = `
Object {
  "events": Object {},
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "key": "/0/div/0/text",
          "text": "Hello,",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "/0/div/1/button/0/text",
              "text": "world",
            },
          ],
          "data": Object {},
          "key": "/0/div/1/button",
          "sel": "button",
        },
        Object {
          "children": Array [],
          "key": "/0/div/2/text",
          "text": "!",
        },
      ],
      "data": Object {},
      "key": "/0/div",
      "sel": "div",
    },
  ],
}
`;

exports[`renders fragments as nested array of elements 1`] = `
Object {
  "events": Object {},
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "key": "/0/h1/0/text",
          "text": "Hello, world!",
        },
      ],
      "data": Object {},
      "key": "/0/h1",
      "sel": "h1",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "key": "/1/h1/0/text",
          "text": "Hello, world!",
        },
      ],
      "data": Object {},
      "key": "/1/h1",
      "sel": "h1",
    },
  ],
}
`;

exports[`renders nested array of elements 1`] = `
Object {
  "events": Object {},
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "key": "/0/h1/0/text",
          "text": "Hello, world!",
        },
      ],
      "data": Object {},
      "key": "/0/h1",
      "sel": "h1",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "key": "/1/h1/0/text",
          "text": "Hello, world!",
        },
      ],
      "data": Object {},
      "key": "/1/h1",
      "sel": "h1",
    },
  ],
}
`;

exports[`renders nested elements 1`] = `
Object {
  "events": Object {},
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "/0/div/0/button/0/text",
              "text": "Red button",
            },
          ],
          "data": Object {
            "style": Object {
              "color": "red",
            },
          },
          "key": "/0/div/0/button",
          "sel": "button",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "/0/div/1/button/0/text",
              "text": "Blue button",
            },
          ],
          "data": Object {
            "style": Object {
              "color": "blue",
            },
          },
          "key": "/0/div/1/button",
          "sel": "button",
        },
      ],
      "data": Object {},
      "key": "/0/div",
      "sel": "div",
    },
  ],
}
`;

exports[`renders single element 1`] = `
Object {
  "events": Object {},
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "key": "/0/button/0/text",
          "text": "Click me!",
        },
      ],
      "data": Object {
        "style": Object {
          "color": "red",
        },
      },
      "key": "/0/button",
      "sel": "button",
    },
  ],
}
`;

exports[`renders styles appending 'px' to number values where possible 1`] = `
Object {
  "events": Object {},
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "key": "/0/div/0/text",
          "text": "Hello, world!",
        },
      ],
      "data": Object {
        "style": Object {
          "flex": 1,
          "height": "50px",
          "width": "100px",
        },
      },
      "key": "/0/div",
      "sel": "div",
    },
  ],
}
`;

exports[`returns options for events 1`] = `
Object {
  "events": Object {
    "/0/input/input": [Function],
  },
  "tree": Array [
    Object {
      "children": Array [],
      "data": Object {
        "attrs": Object {
          "type": "text",
        },
        "on": Object {
          "input": Object {
            "debounce": 500,
          },
        },
        "props": Object {
          "type": "text",
        },
      },
      "key": "/0/input",
      "sel": "input",
    },
  ],
}
`;

exports[`sets SVG attributes 1`] = `
Object {
  "events": Object {},
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "data": Object {
            "attrs": Object {
              "cx": "50",
              "cy": "50",
              "fill": "yellow",
              "r": "20",
              "stroke": "green",
              "stroke-width": "4",
            },
          },
          "key": "/0/svg/0/circle",
          "sel": "circle",
        },
      ],
      "data": Object {
        "attrs": Object {
          "height": "100",
          "width": "100",
        },
      },
      "key": "/0/svg",
      "sel": "svg",
    },
  ],
}
`;

exports[`stores events correctly 1`] = `
Object {
  "events": Object {
    "/0/input/input": [Function],
  },
  "tree": Array [
    Object {
      "children": Array [],
      "data": Object {
        "attrs": Object {
          "type": "text",
        },
        "on": Object {
          "input": true,
        },
        "props": Object {
          "type": "text",
        },
      },
      "key": "/0/input",
      "sel": "input",
    },
  ],
}
`;

exports[`stores events for custom elements correctly 1`] = `
Object {
  "events": Object {
    "/0/custom-element/onInput": [Function],
  },
  "tree": Array [
    Object {
      "children": Array [],
      "data": Object {
        "props": Object {
          "onInput": true,
        },
      },
      "key": "/0/custom-element",
      "sel": "custom-element",
    },
  ],
}
`;
